color/src/cache_key.rs

```diff
@@ -0,0 +1,200 @@
+//!
+//! - Equality on these types is not reflexive (consider [NaN](f32::NAN)).
+//! - Certain values have two representations (`-0` and `+0` are both zero).
+//!
+//! However, it is still useful to create caches which key off these values.
+//! These are caches which don't have any semantic meaning, but instead
+//! are used to avoid redundant calculations or storage.
+//!
+//! Color supports creating these caches by using [`CacheKey<T>`] as the key in
+//! your cache.
+//! `T` is the key type (i.e. a color) which you want to use as the key.
+//! This `T` must implement both [`BitHash`] and [`BitEq`], which are
+//! versions of the standard `Hash` and `Eq` traits which support implementations
+//! for floating point numbers which might be unexpected outside of a caching context.
+
+use core::hash::{Hash, Hasher};
+
+/// A key usable in a hashmap to compare the bit representation
+/// types containing colours.
+///
+/// See the [module level docs](self) for more information.
+#[derive(Debug, Copy, Clone)]
+pub struct CacheKey<T>(pub T);
+
+impl<T> CacheKey<T> {
+    /// Create a new `CacheKey`.
+    ///
+    /// All fields are public, so the struct constructor can also be used.
+    pub fn new(value: T) -> Self {
+        Self(value)
+    }
+
+    /// Get the inner value.
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+// This module exists for these implementations:
+
+// `BitEq` is an equivalence relation, just maybe not the one you'd expect.
+impl<T: BitEq> Eq for CacheKey<T> {}
+impl<T: BitEq> PartialEq for CacheKey<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.bit_eq(&other.0)
+    }
+}
+// If we implement Eq, BitEq's implementation matches that of the hash.
+impl<T: BitHash> Hash for CacheKey<T> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.bit_hash(state);
+    }
+}
+
+/// An hash implementation for types normally wouldn't have one,
+/// implemented using a hash of the bitwise equivalent types when needed.
+///
+/// If a type is `BitHash` and `BitEq`, then it is important that the following property holds:
+///
+/// ```text
+/// k1 biteq k2 -> bithash(k1) == bithash(k2)
+/// ```
+///
+/// See the docs on [`Hash`] for more information.
+///
+/// Useful for creating caches based on exact values.
+/// See the [module level docs](self) for more information.
+pub trait BitHash {
+    /// Feeds this value into the given [`Hasher`].
+    fn bit_hash<H: Hasher>(&self, state: &mut H);
+    // Intentionally no hash_slice for simplicity.
+}
+
+impl BitHash for f32 {
+    fn bit_hash<H: Hasher>(&self, state: &mut H) {
+        self.to_bits().hash(state);
+    }
+}
+impl<T: BitHash, const N: usize> BitHash for [T; N] {
+    fn bit_hash<H: Hasher>(&self, state: &mut H) {
+        self[..].bit_hash(state);
+    }
+}
+
+impl<T: BitHash> BitHash for [T] {
+    fn bit_hash<H: Hasher>(&self, state: &mut H) {
+        state.write_usize(self.len());
```

Comment:
It's not stabilized yet, but when it is, this should use [`Hasher::write_length_prefix`](https://doc.rust-lang.org/stable/std/hash/trait.Hasher.html#method.write_length_prefix)